[docs] Update Config plugins section and add a guide on how to create a config plugin (#35712)
expo/expo: docs/pages/config-plugins/plugins.mdx

---
title: Create and use config plugins
description: Learn how to create and use a config plugins in your Expo project.
sidebar_title: Plugins
---

import { CodeBlocksTable } from '~/components/plugins/CodeBlocksTable';
import { FileTree } from '~/ui/components/FileTree';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

A config plugin is made up of individual plugin functions that modify your app's configuration. These plugin functions have the following characteristics:

> - Plugins are **synchronous** functions that accept an [`ExpoConfig`](/versions/latest/config/app/) and return a modified `ExpoConfig`

Author: > Is it possible to include asynchronous elements inside a config plugin?

Don't know

> Might be a distinction worth noting,...

Do you mean to note here that aync functions are supported but won't be as performant as sync functions? Don't know if "performance" is the right word here, so can you elaborate more on what do you mean by "top-level"?

> - Plugins are **synchronous** functions that accept an [`ExpoConfig`](/versions/latest/config/app/) and return a modified `ExpoConfig`

Author: I'm adding a note about the asynchronous plugin function.

- Plugins should be named using the following convention: `with<Plugin Functionality>`, for example, `withFacebook`
- Plugins should be synchronous and their return value should be serializable, except for any `mods` that are added
- Optionally, a second argument can be passed to the plugin to configure it

> - `plugins` are always invoked when the config is read by the `expo/config` method `getConfig`. However, the `mods` are only invoked during the "syncing" phase of `npx expo prebuild`.

Author: Thanks! I'll rephrase this last point.

    ['plugins/withIosPlugin.ts', 'Contains iOS-specific modifications'],
    ['plugins/withPlugin.ts', 'Main plugin file that combines both Android and iOS plugins'],
    ['app.config.ts', 'App config file that uses the plugin'],
  ]}
/>

<Step label="1">

### Create Android plugin

In **withAndroidPlugin.ts**, add the following code:

```ts withAndroidPlugin.ts
import { ConfigPlugin, withAndroidManifest } from 'expo/config-plugins';

const withAndroidPlugin: ConfigPlugin = config => {
  // Define a custom message
  const message = 'Hello world, from Expo plugin!';

  return withAndroidManifest(config, config => {
    const mainApplication = config?.modResults?.manifest?.application?.[0];

    if (mainApplication) {
      // Ensure meta-data array exists
      if (!mainApplication['meta-data']) {
        mainApplication['meta-data'] = [];
      }

      // Add the custom message as a meta-data entry
      mainApplication['meta-data'].push({
        $: {
          'android:name': 'HelloWorldMessage',
          'android:value': message,
        },
      });
    }

    return config;
  });
};

export default withAndroidPlugin;
```

> The example code above adds a meta-data entry `HelloWorldMessage` to the **android/app/src/main/AndroidManifest.xml** file by importing `ConfigPlugin` and `withAndroidManifest` from `expo/config-plugins` package. The [`withAndroidManifest`](/config-plugins/plugins-and-mods/#what-are-mods) mod is an async function that accepts a config and a data object and modifies the value before returning an object.

Author: Definitions associated with config plugins are going to be listed in the introduction section. I'm working on making that update. This will help us also distill the info on mods page.

</Step>

<Step label="3">

### Create a combined plugin

Now you can create a combined plugin that applies both platform-specific plugins. This approach allows the maintenance of platform-specific code separately while providing a single entry point.

In **withPlugin.ts**, add the following code:

```ts withPlugin.ts
import { ConfigPlugin } from 'expo/config-plugins';
import withAndroidPlugin from './withAndroidPlugin';
import withIosPlugin from './withIosPlugin';

const withPlugin: ConfigPlugin = config => {
  // Apply Android modifications first
  config = withAndroidPlugin(config);
  // Then apply iOS modifications and return
  return withIosPlugin(config);
};

export default withPlugin;
```

</Step>

<Step label="4">

### Add TypeScript support and convert to dynamic app config

> While it is recommended that you write config plugins in TypeScript, as this will provide intellisense for the configuration objects, your app config is ultimately evaluated by Node.js, which does not recognize TypeScript code by default. You need to add a parser for the TypeScript files from the the **plugins** directory to **app.config.ts** file.

Reviewer: ```suggestion
We recommend writing config plugins in TypeScript, since this will provide intellisense for the configuration objects. However, your app config is ultimately evaluated by Node.js, which does not recognize TypeScript code by default. Therefore, you will need to add a parser for the TypeScript files from the the **plugins** directory to **app.config.ts** file.
```
Trying to remove passive voice, but not sure if I used "we" well :)

  return withIosPlugin(config, options);
  /* @end */
};
```

</Step>

<Step label="4">

To pass a value dynamically to the plugin, you can pass an object with the `message` property to the plugin in your app config:

```ts app.config.ts
{
  /* @hide ...*/ /* @end */
  plugins: [
    [
      "./plugins/withPlugin.ts",
      { message: "Custom message from app.config.ts" },
    ],
  ],
}
```

</Step>

## Chaining config plugins

Config plugins can be chained together to apply multiple modifications. Each plugin in the chain runs in the order it appears, with the output of one plugin becoming the input for the next. This sequential execution ensures that dependencies between plugins are respected and allows you to control the precise order of modifications to your native code.

To chain config plugins, you can pass an array of plugins to the `plugins` array property in your app config. This is also supported in JSON app config file format (**app.json**).

```ts app.config.ts
module.exports = ({ config }: { config: ExpoConfig }) => {
  name: 'my app',
  plugins: [
    [withFoo, 'input 1'],
    [withBar, 'input 2'],
    [withDelta, 'input 3'],
  ],
};
```

> The `plugins` array uses `withPlugins` method under the hood to chain the plugins. If you are using so many config plugins with complex configuration that becomes difficult to read and edit your dynamic app config, you can use the `withPlugins` method directly to chain plugins together and execute them in order.

Reviewer: ```suggestion
The `plugins` array uses `withPlugins` method under the hood to chain the plugins. If your plugins array is getting long or has complex configuration, you can use the `withPlugins` method directly to make your configuration easier to read. `withPlugins` will chain the plugins together and execute them in order.
```

```ts app.config.ts
import { withPlugins } from 'expo/config-plugins';

// Create a base config object
const baseConfig = {
  name: 'my app',
  /* @hide ... rest of the config */ /* @end */
};

// ❌ Hard to read
withDelta(withFoo(withBar(config, 'input 1'), 'input 2'), 'input 3');

// ✅ Easy to read
withPlugins(config, [
  [withFoo, 'input 1'],
  [withBar, 'input 2'],
  // When no input is required, you can just pass the method
  withDelta,
]);

// Export the base config with plugins applied
module.exports = ({ config }: { config: ExpoConfig }) => {
  return withPlugins(baseConfig, plugins);
};
```

## Using a config plugin

> Expo config plugins come mostly from Node.js modules. You can install them just like other libraries in your project.

Reviewer: ```suggestion
Expo config plugins are usually included in Node.js modules. You can install them just like other libraries in your project.
```

</Step>

<Step label="3">

### Create a combined plugin

Now you can create a combined plugin that applies both platform-specific plugins. This approach allows the maintenance of platform-specific code separately while providing a single entry point.

In **withPlugin.ts**, add the following code:

```ts withPlugin.ts
import { ConfigPlugin } from 'expo/config-plugins';
import withAndroidPlugin from './withAndroidPlugin';
import withIosPlugin from './withIosPlugin';

const withPlugin: ConfigPlugin = config => {
  // Apply Android modifications first
  config = withAndroidPlugin(config);
  // Then apply iOS modifications and return
  return withIosPlugin(config);
};

export default withPlugin;
```

</Step>

<Step label="4">

### Add TypeScript support and convert to dynamic app config

> While it is convenient to write config plugins in TypeScript, because you can benefit from intellisense on the configuration objects, your app config is evaluated by Node.js, which does not recognize TypeScript by default. You need to add a parser for the TypeScript files from the the **plugins** directory to **app.config.ts** file.

Reviewer: ```suggestion
While it is recommended that you write config plugins in TypeScript, as this will proide intellisense for the configuration objects, your app config is ultimately evaluated by Node.js, which does not recognize TypeScript code by default. You need to add a parser for the TypeScript files from the the **plugins** directory to **app.config.ts** file.
```

  return withIosPlugin(config, options);
  /* @end */
};
```

</Step>

<Step label="4">

To pass a value dynamically to the plugin, you can pass an object with the `message` property to the plugin in your app config:

```ts app.config.ts
{
  /* @hide ...*/ /* @end */
  plugins: [
    [
      "./plugins/withPlugin.ts",
      { message: "Custom message from app.config.ts" },
    ],
  ],
}
```

</Step>

## Chaining config plugins

Config plugins can be chained together to apply multiple modifications. Each plugin in the chain runs in the order it appears, with the output of one plugin becoming the input for the next. This sequential execution ensures that dependencies between plugins are respected and allows you to control the precise order of modifications to your native code.

To chain config plugins, you can pass an array of plugins to the `plugins` array property in your app config. This is also supported in JSON app config file format (**app.json**).

```ts app.config.ts
module.exports = ({ config }: { config: ExpoConfig }) => {
  name: 'my app',
  plugins: [
    [withFoo, 'input 1'],
    [withBar, 'input 2'],
    [withDelta, 'input 3'],
  ],
};
```

> The `plugins` array uses `withPlugins` method under the hood to chain the plugins. In your dynamic app config, if it becomes difficult to read and manipulate, you can use the `withPlugins` method directly to chain plugins together and execute them in order.

Reviewer: ```suggestion
The `plugins` array uses `withPlugins` method under the hood to chain the plugins. If you are using so many config plugins with complex configuration that becomes difficult to read and edit your dynamic app config, you can use the `withPlugins` method directly to chain plugins together and execute them in order.
```

```ts app.config.ts
import { withPlugins } from 'expo/config-plugins';

// Create a base config object
const baseConfig = {
  name: 'my app',
  /* @hide ... rest of the config */ /* @end */
};

// ❌ Hard to read
withDelta(withFoo(withBar(config, 'input 1'), 'input 2'), 'input 3');

// ✅ Easy to read
withPlugins(config, [
  [withFoo, 'input 1'],
  [withBar, 'input 2'],
  // When no input is required, you can just pass the method
  withDelta,
]);

// Export the base config with plugins applied
module.exports = ({ config }: { config: ExpoConfig }) => {
  return withPlugins(baseConfig, plugins);
};
```

## Using a config plugin

> Expo config plugins mostly come from Node.js modules. You can install them just like other libraries in your project.

Reviewer: ```suggestion
Expo config plugins come mostly from Node.js modules. You can install them just like other libraries in your project.
```
